packages/eslint-plugin/tests/rules/no-unnecessary-condition.test.ts

```diff
@@ -150,6 +150,34 @@ function test(a: string | null | undefined) {
 function test(a: unknown) {
   return a ?? "default";
 }`,
+    // Indexing cases
+    `
+declare const arr: object[];
+if(arr[42]) {} // looks unnecessary from the types, but isn't
+
+const tuple = [{}] as [object];
+declare const n: number;
+if(tuple[n]) {}
+`,
+    // Optional-chaining indexing
+    `
+declare const arr: Array<{value: string} & (() => void)>;
+if(arr[42]?.value) {}
+arr[41]?.();
+
+// An array access can "infect" deeper into the chain
+declare const arr2: Array<{x: {y: {z: object}}}>;
+arr2[42]?.x?.y?.z;
+
+const tuple = ["foo"] as const;
+declare const n: number;
+tuple[n]?.toUpperCase();
+    `,
+    // nullish + array index
+    `
+declare const arr: string[][];
+arr[x] ?? [];
```

Comment:
Could you please add the following tests, just for completeness:

```ts
if(arr?.[42]) {}

if(returnsArr?.()[42]) {}

returnsArr?.()[42]?.toUpperCase()
```

Reply:
👍